server/src/main/java/io/crate/auth/JWTAuthenticationMethod.java

```diff
@@ -0,0 +1,178 @@
+        var decodedJWT = credentials.decodedToken();
+        assert decodedJWT != null : "Token must be not null on jwt auth";
+
+        Role user = roles.findUser(username);
+        if (user == null) {
+            throw new RuntimeException("jwt authentication failed for user \"" + username + "\"");
+        }
+
+        try {
+            JwkProvider jwkProvider = jwkProvider(decodedJWT);
+            Algorithm algorithm = resolveAlgorithm(jwkProvider, decodedJWT);
+
+            var jwtProperties = user.jwtProperties();
+            assert jwtProperties != null : "Resolved user must have jwt properties";
+            // Expiration date is checked by default(if provided in token)
+            JWTVerifier.BaseVerification verification = (JWTVerifier.BaseVerification) JWT.require(algorithm)
+                // withers below are not needed for payload signature check.
+                // It's an extra step on top of signature verification to double check that user metadata matches token payload.
+                .withIssuer(jwtProperties.iss())
+                .withClaim("username", jwtProperties.username());
+
+            JWTVerifier verifier = verification.build(clock());
+            verifier.verify(decodedJWT);
+
+        } catch (Exception e) {
+            throw new RuntimeException(
+                String.format(
+                    Locale.ENGLISH,
+                    "jwt authentication failed for user %s. Reason: %s",
+                    username,
+                    e.getMessage()
+                )
+            );
+        }
+
+        return user;
+    }
+
+    public Clock clock() {
+        return Clock.systemUTC();
+    }
+
+    public JwkProvider jwkProvider(@NotNull DecodedJWT decodedJWT) {
+        URL jwkUrl;
+        try {
+            /*
+             We cannot use JwkProviderBuilder constructor with String.
+             It adds hard coded WELL_KNOWN_JWKS_PATH (/.well-known/jwks.json) to the url.
+             JWK URL not necessarily ends with that suffix, for example:
+             Microsoft: "jwks_uri":" https://login.microsoftonline.com/common/discovery/v2.0/keys"
+             Google: "jwks_uri": "https://www.googleapis.com/oauth2/v3/certs",
+             Taken from https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration
+             and https://accounts.google.com/.well-known/openid-configuration correspondingly
+            */
+
+            URI uri = new URI(decodedJWT.getIssuer()).normalize();
+            jwkUrl = uri.toURL();
+        } catch (MalformedURLException | URISyntaxException e) {
+            throw new IllegalArgumentException("Invalid jwks uri", e);
+        }
+
+        return new JwkProviderBuilder(jwkUrl)
+            // up to 10 JWKs will be cached for up to 24 hours
+            .cached(10, 24, TimeUnit.HOURS)
+            // up to 10 JWKs can be retrieved within one minute
+            .rateLimited(10, 1, TimeUnit.MINUTES)
```

Comment:
Can we comment the why instead of the what? (what is not really needed, the function call speak for itself).
I'd also suggest to define these values statically to make them more visible.

Reply:
I will use default values instead https://github.com/crate/crate/pull/15551#discussion_r1497298904